crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+            self.delegates.clone()
+        };
+        for delegate in lock_free_calls.as_slice() {
+            if let Err(error) = delegate.call(&mut callback) {
+                const RPC_E_SERVER_UNAVAILABLE: HRESULT = HRESULT(-2147023174); // HRESULT_FROM_WIN32(RPC_S_SERVER_UNAVAILABLE)
+                if matches!(error.code(), RPC_E_DISCONNECTED | JSCRIPT_E_CANTEXECUTE | RPC_E_SERVER_UNAVAILABLE) {
+                    self.remove(delegate.to_token())?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// A thread-safe reference-counted array of delegates.
+struct Array<T: Interface + Clone> {
+    buffer: *mut Buffer,
+    len: usize,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: Interface + Clone> Default for Array<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Interface + Clone> Array<T> {
+    /// Creates a new, empty `Array<T>` with no capacity.
+    fn new() -> Self {
+        Self { buffer: std::ptr::null_mut(), len: 0, _phantom: std::marker::PhantomData }
+    }
+    /// Creates a new, empty `Array<T>` with the specified capacity.
+    fn with_capacity(capacity: usize) -> Result<Self> {
+        Ok(Self { buffer: Buffer::new(capacity * std::mem::size_of::<Delegate<T>>())?, len: 0, _phantom: std::marker::PhantomData })
+    }
+    /// Swaps the contents of two `Array<T>` objects.
+    fn swap(&mut self, mut other: Self) -> Self {
+        unsafe { std::ptr::swap(&mut self.buffer, &mut other.buffer) };
+        std::mem::swap(&mut self.len, &mut other.len);
+        other
+    }
+    /// Returns `true` if the array contains no delegates.
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+    /// Returns the number of delegates in the array.
+    fn len(&self) -> usize {
+        self.len
+    }
+    /// Appends a delegate to the back of the array.
+    fn push(&mut self, delegate: Delegate<T>) {
+        unsafe {
+            std::ptr::write((*self.buffer).as_mut_ptr::<Delegate<T>>().add(self.len) as _, delegate);
+            self.len += 1;
+        }
+    }
+    /// Returns a slice containing of all delegates.
+    fn as_slice(&self) -> &[Delegate<T>] {
+        if self.is_empty() {
+            &[]
+        } else {
+            unsafe { std::slice::from_raw_parts((*self.buffer).as_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+    /// Returns a mutable slice of all delegates.
+    fn as_mut_slice(&mut self) -> &mut [Delegate<T>] {
+        if self.is_empty() {
+            &mut []
+        } else {
+            unsafe { std::slice::from_raw_parts_mut((*self.buffer).as_mut_ptr::<Delegate<T>>() as _, self.len) }
+        }
+    }
+}
+
+impl<T: Interface + Clone> Clone for Array<T> {
+    fn clone(&self) -> Self {
+        if !self.is_empty() {
+            unsafe { (*self.buffer).0.add_ref() };
+        }
+        Self { buffer: self.buffer, len: self.len, _phantom: std::marker::PhantomData }
```

Comment:
If I clone this `Array` I then have two `Array<T>` which I can call `push` on at the same time. However, if they are pointing to the same buffer, I might be overwriting their contents. This would seem to violate memory safety, no?

Reply:
This is just bumping the ref count on the same array. Basically, there are now two smart pointers to the same array to support lock-free calling. The `swap`/`change` locks ensures we don't touch the array being used for calls.